controller/controller_server/csi_controller_server.py

```diff
@@ -534,6 +539,81 @@ def GetCapacity(self, request, context):
         logger.info("finished GetCapacity")
         return csi_pb2.GetCapacityResponse()
 
+    def ControllerExpandVolume(self, request, context):
+        set_current_thread_name(request.volume_id)
+        logger.info("ControllerExpandVolume")
+        secrets = request.secrets
+
+        try:
+            utils.validate_expand_volume_request(request)
+
+            user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+
+            try:
+                array_type, volume_id = utils.get_volume_id_info(request.volume_id)
+            except ObjectIdError as ex:
+                logger.warning("volume id is invalid. error : {}".format(ex))
+                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                return csi_pb2.ControllerExpandVolumeResponse()
+
+            with get_agent(user, password, array_addresses, array_type).get_mediator() as array_mediator:
+                logger.debug(array_mediator)
+
+                size = request.capacity_range.required_bytes
+
+                if size > array_mediator.maximal_volume_size_in_bytes:
+                    message = messages.SizeOutOfRangeError_message.format(size,
+                                                                          array_mediator.maximal_volume_size_in_bytes)
+                    context.set_details(message)
+                    context.set_code(grpc.StatusCode.OUT_OF_RANGE)
+                    return csi_pb2.ControllerExpandVolumeResponse()
+
+                logger.debug("expanding volume {0}".format(volume_id))
+                array_mediator.expand_volume(
+                    volume_id=volume_id,
+                    required_bytes=size)
+
+                volume = array_mediator.get_object_by_id(volume_id, config.VOLUME_TYPE_NAME)
```

Comment:
wouldn't it be more efficient if `expand_volume` would just return the volume (as was initially documented in the mediator interface)?

Reply:
the mediators never got them.. at least not all of them

Reply:
then wouldn't it be more efficient for those who did get them?

Reply:
done (working on mediators now)